Lib/compression/zstd/_zstdfile.py

```diff
@@ -0,0 +1,378 @@
+        an existing file object to read from or write to.
+
+        mode can be "r" for reading (default), "w" for (over)writing, "x" for
+        creating exclusively, or "a" for appending. These can equivalently be
+        given as "rb", "wb", "xb" and "ab" respectively.
+
+        Parameters
+        level: The compression level to use, defaults to ZSTD_CLEVEL_DEFAULT. Note,
+            in read mode (decompression), compression level is not supported.
+        options: A dict object, containing advanced compression
+            parameters.
+        zstd_dict: A ZstdDict object, pre-trained dictionary for compression /
+            decompression.
+        """
+        self._fp = None
+        self._closefp = False
+        self._mode = _MODE_CLOSED
+
+        # Read or write mode
+        if mode in ("r", "rb"):
+            if not isinstance(options, (type(None), dict)):
+                raise TypeError(
+                    (
+                        "In read mode (decompression), options argument "
+                        "should be a dict object, that represents decompression "
+                        "options."
+                    )
+                )
+            if level is not None:
+                raise TypeError("level argument should only be passed when writing.")
+            mode_code = _MODE_READ
+        elif mode in ("w", "wb", "a", "ab", "x", "xb"):
+            if not isinstance(level, (type(None), int)):
+                raise TypeError("level argument should be an int object.")
+            if not isinstance(options, (type(None), dict)):
+                raise TypeError("options argument should be an dict object.")
+            mode_code = _MODE_WRITE
+            self._compressor = ZstdCompressor(
+                level=level, options=options, zstd_dict=zstd_dict
+            )
+            self._pos = 0
+        else:
+            raise ValueError(f"Invalid mode: {mode!r}")
+
+        # File object
+        if isinstance(filename, (str, bytes, PathLike)):
+            if "b" not in mode:
+                mode += "b"
+            self._fp = builtins.open(filename, mode)
+            self._closefp = True
+        elif hasattr(filename, "read") or hasattr(filename, "write"):
+            self._fp = filename
+        else:
+            raise TypeError("filename must be a str, bytes, file or PathLike object")
+        self._mode = mode_code
+
+        if self._mode == _MODE_READ:
+            raw = self._READER_CLASS(
+                self._fp,
+                ZstdDecompressor,
+                trailing_error=ZstdError,
+                zstd_dict=zstd_dict,
+                options=options,
+            )
+            self._buffer = io.BufferedReader(raw)
+
+    def close(self):
+        """Flush and close the file.
+
+        May be called more than once without error. Once the file is
+        closed, any other operation on it will raise a ValueError.
+        """
+        # Nop if already closed
```

Comment:
prefer either capitalised NOP or fully spelt out, but in this case I think the comment can be removed
```suggestion
```